### PR TITLE
Federation Sender

### PR DESCRIPTION
**Pull request by Half-Shot**
This now supports:
- [x] EDUs 
  - [x] Regular
  - [x] Keyed
  - [x] Device Messaging
  - [x] Device Lists (basic, no use of `prev_ids` yet)
- [x] PDUs 
- [x] Metrics

More things that need to be done:
- [x] Well-known
- [x] Basic SRV support
- [ ] SRV load balancing / failover
- [x] Connect via HTTPS
   - [x] Support self signed certs optionally
- [ ] Get hosts in rooms should be not hacky
- [x] Ack presence

Perf issues/ Bugs:

- [ ] On startup we treat all hosts equally, so dead hosts get as much time as live ones. We should use the `destinations` table and refine how we backoff.
- [ ] Presence still fails to sign for ~30% of requests. This isn't a huge issue but it's annoying.
- [ ] On a "large" host like half-shot.uk, the CPU can go haywire if you don't limit the tx concurrency, and it lags by minutes if you do.
- [ ] Floods of DNS requests via well-known (and generally sending a lot of them) makes things very sad.

### CODE REVIEW

**Comment by Half-Shot** [comment]
This now works for EDUs, and will work for PDUs

**Comment by Half-Shot** [comment]
I've added support for metrics, which is probably very naughty on this branch but I wanted to measure performance.

File: Matrix.SynapseInterop.Common/WorkerMetrics.cs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    public static class WorkerMetrics
+    {
+        private const string PREFIX = "synapse_netcore_worker";
+
+        static readonly Counter TransactionsSent = Metrics.CreateCounter(
+            $"{PREFIX}_txns_sent",
+            "Number of transactions sent",
+            new CounterConfiguration
+            {
+                LabelNames = new[] { "instance", "outcome", "destination" }
```
Comment (turt2live) [inline Matrix.SynapseInterop.Common/WorkerMetrics.cs]: lint that doesn't exist: trailing comma

File: Matrix.SynapseInterop.Common/WorkerMetrics.cs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+            }
+        );
+
+        static readonly Histogram TransactionDuration = Metrics.CreateHistogram(
+            $"{PREFIX}_txns_duration",
+            "Time taken to complete a transaction",
+            new HistogramConfiguration
+            {
+                LabelNames = new[] { "instance", "destination" }
+
```
Comment (turt2live) [inline Matrix.SynapseInterop.Common/WorkerMetrics.cs]: extreme nit: blank space

File: Matrix.SynapseInterop.Common/WorkerMetrics.cs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+                LabelNames = new[] { "instance", "outcome", "destination" }
+            }
+        );
+
+        static readonly Histogram TransactionDuration = Metrics.CreateHistogram(
+            $"{PREFIX}_txns_duration",
+            "Time taken to complete a transaction",
+            new HistogramConfiguration
+            {
+                LabelNames = new[] { "instance", "destination" }
```
Comment (turt2live) [inline Matrix.SynapseInterop.Common/WorkerMetrics.cs]: lint that doesn't exist: trailing comma

File: Matrix.SynapseInterop.Common/WorkerMetrics.cs [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+                
+            }
+        );
+        private static MetricServer _srv;
+        private static string _name;
+        public static void StartMetrics (string bindHost, int bindPort, string instanceName)
```
Comment (turt2live) [inline Matrix.SynapseInterop.Common/WorkerMetrics.cs]: lint that doesn't exist: blank line between variables and functions

File: Matrix.SynapseInterop.Database/SynapseDbContext.cs [hunk 5]
```diff
@@ -18,5 +31,52 @@ protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
...
             optionsBuilder.UseNpgsql(_connString);
         }
+
+        public List<DeviceContentSet> GetNewDevicesForDestination(string destination, int limit)
+        {
+            List<DeviceContentSet> set = new List<DeviceContentSet>();
+            foreach (var devicePoke in DeviceListsOutboundPokes
+                .Where(poke => poke.Destination == destination && poke.Sent == false)
+                .Take(limit)
+                .OrderBy(poke => poke.StreamId).ToList())
```
Comment (turt2live) [inline Matrix.SynapseInterop.Database/SynapseDbContext.cs]: lint that doesn't exist: linq queries should be assigned to variables

File: Matrix.SynapseInterop.Database/SynapseDbContext.cs [hunk 6]
```diff
@@ -18,5 +31,52 @@ protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
...
+            return set;
+        }
+
+        public List<EventJsonSet> GetAllNewEventsStream(int fromId, int currentId, int limit)
+        {
+            List<EventJsonSet> set = new List<EventJsonSet>();
+            foreach (var ev in Events
+                .Where(e => e.StreamOrdering > fromId && e.StreamOrdering <= currentId)
+                .Take(limit)
+                .OrderBy((e) => e.StreamOrdering).ToList())
```
Comment (turt2live) [inline Matrix.SynapseInterop.Database/SynapseDbContext.cs]: lint that doesn't exist: linq queries should be assigned to variables

File: Matrix.SynapseInterop.Database/SynapseDbContext.cs [hunk 7]
```diff
@@ -18,5 +31,52 @@ protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
         {
             optionsBuilder.UseNpgsql(_connString);
         }
+
+        public List<DeviceContentSet> GetNewDevicesForDestination(string destination, int limit)
```
Comment (turt2live) [inline Matrix.SynapseInterop.Database/SynapseDbContext.cs:36]: lint that doesn't exist: `IEnumberable` should be exposed, not a mutable collection

File: Matrix.SynapseInterop.Database/SynapseDbContext.cs [hunk 8]
```diff
@@ -18,5 +31,52 @@ protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
...
+                contentSet.keys = JObject.Parse(json.KeyJson);
+                contentSet.device_display_name = Devices.SingleOrDefault(dev =>
+                    dev.UserId == devicePoke.UserId && dev.DeviceId == devicePoke.DeviceId)
+                    ?.DisplayName;
+                set.Add(contentSet);
+            }
+            return set;
+        }
+
+        public List<EventJsonSet> GetAllNewEventsStream(int fromId, int currentId, int limit)
```
Comment (turt2live) [inline Matrix.SynapseInterop.Database/SynapseDbContext.cs:73]: lint that doesn't exist: `IEnumberable` should be exposed, not a mutable collection

File: Matrix.SynapseInterop.Database/SynapseDbContext.cs [hunk 9]
```diff
@@ -18,5 +31,52 @@ protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
...
+                .OrderBy((e) => e.StreamOrdering).ToList())
+            {
+                var js = EventsJson.SingleOrDefault(e => e.EventId == ev.EventId);
+                if (js != null)
+                    set.Add(
+                        new EventJsonSet(
+                            ev,
+                            js.Json,
+                            js.FormatVersion
+                        )
```
Comment (turt2live) [inline Matrix.SynapseInterop.Database/SynapseDbContext.cs]: lint that doesn't exist: this is a one-liner

File: Matrix.SynapseInterop.Replication/DataRows/FederationStreamRow.cs [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+        public List<EduEvent> edus;
+        public List<string> devices;
+        public List<PresenceState> presence;
+
+        private FederationStreamRow()
+        {
+            keyedEdus = new Dictionary<string[], EduEvent>();
+            edus = new List<EduEvent>();
+            devices = new List<string>();
+            presence = new List<PresenceState>();
```
Comment (turt2live) [inline Matrix.SynapseInterop.Replication/DataRows/FederationStreamRow.cs:21]: lint that doesn't exist: assign variables where they are defined, not in the constructor

File: Matrix.SynapseInterop.Replication/DataRows/FederationStreamRow.cs [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+        {
+            var parsed = JsonConvert.DeserializeObject<List<dynamic>>(rawDataString);
+            string typeId = "";
+            FederationStreamRow streamRow = new FederationStreamRow();
+            for (int i = 0; i < parsed.Count; i++)
+            {
+                if (i == 0)
+                {
+                    typeId = parsed[0];
+                } else if (typeId == "k") // KeyedEduRow
```
Comment (turt2live) [inline Matrix.SynapseInterop.Replication/DataRows/FederationStreamRow.cs]: lint that doesn't exist: else if on a dedicated line

x4

File: Matrix.SynapseInterop.Replication/SynapseReplication.cs [hunk 12]
```diff
@@ -76,7 +76,14 @@ private void ReadLoop()
...
                 } while (stream.DataAvailable);
 
-                ProcessCommands(result.ToString());
+                try
+                {
+                    ProcessCommands(result.ToString());
+                }
+                catch (Exception ex)
+                {
+                    Console.WriteLine("Failed to process command: {0}", ex);
```
Comment (turt2live) [inline Matrix.SynapseInterop.Replication/SynapseReplication.cs:101]: We should absolutely not be catching here unless we're going to write an `ERROR` over the line and close the connection.

File: Matrix.SynapseInterop.Worker.AppserviceSender/Controllers/AppserviceAdminRouting.cs [hunk 13]
```diff
@@ -0,0 +1,40 @@
+﻿using Newtonsoft.Json.Linq;
+using Routable;
+using Routable.Kestrel;
+
+namespace Matrix.SynapseInterop.Worker.AppserviceSender.Controllers
+{
+    public sealed class AppserviceAdminRouting : KestrelRouting
```
Comment (turt2live) [inline Matrix.SynapseInterop.Worker.AppserviceSender/Controllers/AppserviceAdminRouting.cs]: err... probably want to rebase the appservice crap out of this

File: Matrix.SynapseInterop.Worker.FederationSender/FederationClient.cs [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+using Newtonsoft.Json.Linq;
+
+namespace Matrix.SynapseInterop.Worker.FederationSender
+{
+    class FederationClient
+    {
+        private SigningKey key;
+        private HttpClient client;
+        private string origin;
+        public FederationClient(string serverName, SigningKey key)
```
Comment (turt2live) [inline Matrix.SynapseInterop.Worker.FederationSender/FederationClient.cs]: lint that doesn't exist: newline between variables and methods

File: Matrix.SynapseInterop.Worker.FederationSender/FederationClient.cs [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+        {
+            origin = serverName;
+            client = new HttpClient();
```
Comment (turt2live) [inline Matrix.SynapseInterop.Worker.FederationSender/FederationClient.cs]: lint that doesn't exist: assign variables where they are defined, not in the constructor

File: Matrix.SynapseInterop.Worker.FederationSender/FederationClient.cs [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+
+        public async Task SendTransaction(Transaction transaction)
+        {
+            var uri = new UriBuilder(GetUrlForDestination(transaction.destination));
+            uri.Path += $"send/{transaction.transaction_id}/";
+            Console.WriteLine($"[TX] PUT {uri} ");
+            var msg = new HttpRequestMessage
+            {
+                Method = HttpMethod.Put,
+                RequestUri = uri.Uri
```
Comment (turt2live) [inline Matrix.SynapseInterop.Worker.FederationSender/FederationClient.cs]: lint that doesn't exist: trailing comma

File: Matrix.SynapseInterop.Worker.FederationSender/FederationClient.cs [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+            };
+            var body = SigningKey.SortPropertiesAlphabetically(JObject.FromObject(transaction));
+            SignRequest(msg, transaction.destination, body);
+            var json = JsonConvert.SerializeObject(body, Formatting.None);
+            var content = new StringContent(
+                json,
+                Encoding.UTF8,
+                "application/json"
```
Comment (turt2live) [inline Matrix.SynapseInterop.Worker.FederationSender/FederationClient.cs]: lint that doesn't exist: this is a one-liner

File: Matrix.SynapseInterop.Worker.FederationSender/FederationClient.cs [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+            );
+            msg.Content = content;
+            var resp = await client.SendAsync(msg);
+            Console.WriteLine($"[TX] Response: {resp.StatusCode} {resp.ReasonPhrase}");
+            if (resp.IsSuccessStatusCode)
+            {
+                return;
```
Comment (turt2live) [inline Matrix.SynapseInterop.Worker.FederationSender/FederationClient.cs:120]: lint that doesn't exist: inline this

File: Matrix.SynapseInterop.Worker.FederationSender/FederationClient.cs [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+
+        private void SignRequest(HttpRequestMessage msg, string destination, JObject body = null)
+        {
+            var sigBody = new JObject();
+            sigBody.Add(origin, new JObject());
+            sigBody[origin].Value<JObject>().Add($"{key.Type}:${key.Name}", key.PublicKey);
+            var signingBody = new JObject();
+            if (body != null)
+            {
+                signingBody.Add("content", body);
```
Comment (turt2live) [inline Matrix.SynapseInterop.Worker.FederationSender/FederationClient.cs:167]: lint that doesn't exist: inline this

File: Matrix.SynapseInterop.Worker.FederationSender/FederationSender.cs [hunk 20]
```diff
@@ -0,0 +1,134 @@
...
+            UpdateFederationPos("federation", _stream_position);
+            _synapseReplication.SendFederationAck(_stream_position.ToString());
+            _last_ack = token;
+        }
+
+        private void ProcessEventQueueLoop()
+        {
+            while (true) // This will be broken out of at some point.
+            {
+
```
Comment (turt2live) [inline Matrix.SynapseInterop.Worker.FederationSender/FederationSender.cs]: TODO please

File: Matrix.SynapseInterop.Worker.FederationSender/SigningKey.cs [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+                throw new InvalidDataException("Key file is in the wrong format. Should be '$type $name $key'");
+            }
+
+            Type = split[0];
+            Name = split[1];
+            // Synapse uses unpadded base64, so we need to add our own padding to make this work.
+            var b64Seed = split[2].Trim();
+            while (b64Seed.Length % 4 != 0)
+            {
+                b64Seed += "=";
```
Comment (turt2live) [inline Matrix.SynapseInterop.Worker.FederationSender/SigningKey.cs:39]: lint that doesn't exist: inline this

File: Matrix.SynapseInterop.Worker.FederationSender/SigningKey.cs [hunk 22]
```diff
@@ -0,0 +1,99 @@
...
+            var enumerable = original.Properties().ToList().OrderByOrdinal(p => p.Name);
+            foreach (var property in enumerable)
+            {
+                var value = property.Value as JObject;
+                var valueArr = property.Value as JArray;
+                if (value != null)
+                {
+                    value = SortPropertiesAlphabetically(value);
+                    result.Add(property.Name, value);
+                } else if (valueArr != null)
```
Comment (turt2live) [inline Matrix.SynapseInterop.Worker.FederationSender/SigningKey.cs]: lint that doesn't exist: else if on a dedicated line

File: Matrix.SynapseInterop.Worker.FederationSender/SigningKey.cs [hunk 23]
```diff
@@ -0,0 +1,99 @@
...
+                {
+                    var res = new JArray();
+                    foreach (var jToken in valueArr)
+                    {
+                        if (jToken is JObject o)
+                        {
+                            res.Add(SortPropertiesAlphabetically(o));
```
Comment (turt2live) [inline Matrix.SynapseInterop.Worker.FederationSender/SigningKey.cs:86]: lint that doesn't exist: inline these:
```
if (condition) doThing();
else doOtherThing();
```

File: Matrix.SynapseInterop.Worker.FederationSender/TransactionQueue.cs [hunk 24]
```diff
@@ -0,0 +1,507 @@
...
+        private string connString;
+        private int txnId;
+        private int lastEventPoke;
+
+        private Dictionary<string, long> destLastDeviceMsgStreamId;
+        private Dictionary<string, long> destLastDeviceListStreamId;
+
+        public TransactionQueue(string serverName, string connectionString, SigningKey key)
+        {
+            client = new FederationClient(serverName, key);
```
Comment (turt2live) [inline Matrix.SynapseInterop.Worker.FederationSender/TransactionQueue.cs]: lint that doesn't exist: assign variables where they are defined, not in the constructor

File: Matrix.SynapseInterop.Worker.FederationSender/TransactionQueue.cs [hunk 25]
```diff
@@ -0,0 +1,507 @@
...
+            lastEventPoke = -1;
+            signingKey = key;
+        }
+
+        public void OnEventUpdate(string stream_pos)
+        {
+            lastEventPoke = Math.Max(int.Parse(stream_pos), lastEventPoke);
+            if (!eventsProcessing.IsCompleted)
+            {
+                return;
```
Comment (turt2live) [inline Matrix.SynapseInterop.Worker.FederationSender/TransactionQueue.cs:74]: lint that doesn't exist: inline this

File: Matrix.SynapseInterop.Worker.FederationSender/TransactionQueue.cs [hunk 26]
```diff
@@ -0,0 +1,507 @@
...
+                    {
+                        userPresence.Remove(presence.user_id);
+                        userPresence.Add(presence.user_id, presence);
+                    }
+                }
+            }
+
+            if (!presenceProcessing.IsCompleted)
+            {
+                return;
```
Comment (turt2live) [inline Matrix.SynapseInterop.Worker.FederationSender/TransactionQueue.cs:105]: lint that doesn't exist: inline this

xLots

**Review by turt2live — APPROVED** [review]
Generally looks sane to me as C# code, despite my super picky linting preferences (https://github.com/turt2live/synapse-netcore-workers/issues/10)

I'm generally fine if you want to ignore all of the linting for now and I'll deal with it later - there are a couple other nits buried in this review somewhere.

**Review by turt2live — APPROVED** [review]
You're trustworthy or something

**Comment by Half-Shot** [comment]
It builds

Ship it

**Pull request merged**